File: provider/money24_rate_provider.py
from typing import Optional

from client.money24_client import Money24Client
from provider.base.rate_provider import RateProvider
from statics import types


class Money24RateProvider(RateProvider):
    def __init__(self, money24_client: Money24Client):
        self._money24_client = money24_client

    async def get_rates(self) -> Optional[dict]:
        rates = await self._money24_client.get_rates()
        result = {}
        for rate in rates['result']:
            currency = str(rate['currCode']).upper()
            typed_rate = result.setdefault(currency, {})
            typed_rate[types.BUY if rate['type'] == 'buy' else types.SELL] = rate['rate']
            rate[currency] = typed_rate
        return result

    async def save_rates(self, rates: dict):
        raise NotImplementedError()

File: src/main/java/com/ua/money24/service/provider/rate/Money24RateProvider.java
package com.ua.money24.service.provider.rate;

import com.ua.money24.client.Money24Client;
import com.ua.money24.constants.ExecTypes;
import com.ua.money24.constants.Methods;
import com.ua.money24.constants.RateType;
import com.ua.money24.exceptions.RatesNotFoundException;
import com.ua.money24.model.Rate;
import com.ua.money24.model.request.ExecAsPublicRequest;
import com.ua.money24.model.response.ExecAsPublicResponse;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

@Component
public class Money24RateProvider implements RateProvider {
    private final Money24Client money24Client;

    public Money24RateProvider(Money24Client money24Client) {
        this.money24Client = money24Client;
    }

    @Override
    public List<Rate> getRatesInRegion(int regionId) {
        var response = money24Client.execAsPublic(new ExecAsPublicRequest(
                ExecTypes.GET,
                String.format(Methods.RATES, regionId),
                null
        ));
        if (response.statusCode() != HttpStatus.OK.value())
            throw new RatesNotFoundException();
        return response.result()
                .rates()
                .stream()
                .filter(rate -> rate.rate() > 0)
                .collect(Collectors.groupingBy(ExecAsPublicResponse.Result.Rate::id))
                .entrySet()
                .stream()
                .map(entry -> {
                    int id = entry.getKey();
                    var buyRate = findRateByType(entry.getValue(), RateType.BUY);
                    var sellRate = findRateByType(entry.getValue(), RateType.SELL);
                    return new Rate(
                            id,
                            buyRate.regionID(),
                            buyRate.currCode(),
                            buyRate.currId(),
                            buyRate.rate(),
                            sellRate.rate()
                    );
                })
                .collect(Collectors.toList());
    }

    private ExecAsPublicResponse.Result.Rate findRateByType(List<ExecAsPublicResponse.Result.Rate> rates, String type) {
        return rates.stream()
                .filter(rate -> type.equals(rate.type()))
                .findFirst().orElse(null);
    }
}
